### DISCUSSION TITLE
Memory leak with Nuxt3

### DISCUSSION THREAD

**Discussion by jeanmatthieud** [nuxt/nuxt#23553] [Questions] [answered]
Hi 👋 

I recently migrated an app from Nuxt2 to Nuxt3, rewriting most of the components, pages and Nuxt configuration.

Now, it works fine ! However, after deploying it on production (and fixing small bugs), I discovered some kind of memory leak:
![image](https://github.com/nuxt/nuxt/assets/1443167/10070484-bc8e-43fb-a82b-32c29ab44001)
(green arrow: manual deployment / reboot)

The server respond to about 100-500 requests per minute (avg 200 per minute).

I tried to find the issue with Node Chrome Profiler + [k6](https://k6.io/), and [Clinic](https://clinicjs.org/heapprofiler/) + autocannon, without finding the issue.

I'm stuck ! Maybe it is on a component which was not on the pages visited by the tools...  
I don't use the "server" folder, except one nitro plugin to redirect some requests.  

What I tried (without fixing the issue):
- I removed the manual implementation of Sentry
- I checked that every ref / reactive / watch (etc...) is called in a 'setup()' function (or in `<script setup>`)
- I checked that no plugin is storing data in a global var
- I checked that my code is not calling any Vue.something()

I'm using Node 20 (tried Node 18), and the following packages:
```json
"devDependencies": {
    "@nuxt/devtools": "latest",
    "@nuxtjs/mdc": "^0.2.1",
    "@nuxtjs/tailwindcss": "^6.8.0",
    "@pinia-plugin-persistedstate/nuxt": "^1.1.2",
    "@tailwindcss/forms": "^0.5.6",
    "@types/lodash": "^4.14.199",
    "@types/qs": "^6.9.8",
    "@zadigetvoltaire/nuxt-gtm": "^0.0.13",
    "nuxt": "^3.7.4"
  },
  "dependencies": {
    "@nuxt/content": "^2.8.5",
    "@nuxt/image": "^1.0.0-rc.2",
    "@pinia/nuxt": "^0.4.11",
    "@vee-validate/nuxt": "^4.11.7",
    "@vee-validate/yup": "^4.11.7",
    "awesome-phonenumber": "^6.0.0",
    "bignumber.js": "^9.1.2",
    "dayjs": "^1.11.10",
    "keen-slider": "^6.8.6",
    "ky": "^1.0.1",
    "ky-universal": "^0.12.0",
    "lodash-es": "^4.17.21",
    "medium-zoom": "^1.0.8",
    "mitt": "^3.0.1",
    "nuxt-jsonld": "^2.0.8",
    "pinia": "^2.1.6",
    "qs": "^6.11.2",
    "swrv": "^1.0.4",
    "yup": "^1.3.2",
    "yup-locales": "^1.2.18"
  },
```

I'm really stuck... I cannot share the source code (it's a proprietary website).  
Please help me 😢

**Comment by huang-julien** [comment]
Hi :wave: if you can't share the code, the only solution would be to un-scope (remove) fucntionnalities/features of your website until finding the source.

Make sure you don't have any `setInterval`/`setTimeout` server-side , correctly await promises and not assigning app data to shared objects.

**Reply by jeanmatthieud** [reply]
Thanks

It's difficult to remove functionalities on an app in production. I've done it for non-essential functionalities, but I cannot remove the pages...

I checked my code, and I don't have any `setInterval` or `setTimeout`, and I don't think I'm assigning app data to shared object.
I'm checking the promises, but so far, I don't see where it could be a problem.

**Reply by huang-julien** [reply]
You can also try getting a  memory dump to analyze what objects are in the global scope (even if it's quite complicated to read, thank you javascript)

hmmm... you should ask a dev instance to deploy a build and do some load testing on it (you can do it with K6 for example), or even do it locally

**Reply by jeanmatthieud** [reply]
I will try to build a dev instance, but I'm afraid that the memory leak doesn't occurs with basic navigation scenarios.

Is it possible that the memory is not garbage collected before the crash ?

**Comment by jeanmatthieud** [comment]
Hi !
I narrowed it down to the use of `ky-universal` to call the API.  
I mocked the API calls, and the memory leak disappeared.  
Then, I did a quick and dirty replacement of `ky` by `ofetch` (but without configuring the HTTP hooks). And no issue...

Some load tests with k6:

![image](https://github.com/nuxt/nuxt/assets/1443167/3ec3c503-133b-42b9-9f67-afcf5fcedd07)

- In red: Memory leaks, removing/mocking functionalities of the project step by step
- In orange: API calls mocked, lot of functionalities removed, no memory leak (same load test)
- In green: All functionalities restored, `ky-universal` removed, most API calls done with `ofetch` (dirty wrapping)

It was working fine with nuxt2. Each `GET` call is wrapped in a `useAsyncData`.  
I will try to check if a specific hook of `ky-universal` is misbehaving or if I have to move all calls to `ofetch` (but requires a lot of rework and tests).

**Comment by jeanmatthieud** [comment, marked as answer]
I finally found the issue.  
I was registering a `ky` `afterResponse` hook:
```js
const $apiHttp = ky.extend({
  prefixUrl: nuxtApp.$config.public.apiUrl,
  hooks: {
    afterResponse: [
      async (input, options, response) => {
        if (['GET', 'POST'].includes(input.method) && response.status === 401 && otherCheckSnipped) {
          console.info(`❌ User auth problematic for cart fetching, disconnect, remove auth and retry`);
          // Disconnect
          await clearAuth();
          // Remove auth token
          const reqHeaders = new Headers(options.headers);
          reqHeaders.delete('Authorization');
          options.headers = reqHeaders;
          options.retry.limit = 0;
          return input.method === "GET" ? ky.get(input, options) : ky.post(input, options);
        }
        // 🎉 The memory leak disappears after adding this line
        return response;
      },
    ],
  }
});
```

This is odd, it was working fine with Nuxt2.  Moreover, the official [Ky documentation](https://github.com/sindresorhus/ky#hooksafterresponse) doesn't return a response.
I'm quite happy to close this issue and move on!

Production server:
![image](https://github.com/nuxt/nuxt/assets/1443167/3605d061-d7c3-4602-ae5f-b1dd2d5af6ae)

**Reply by jeanmatthieud** [reply]
I also discovered that ky-universal is not required since ky@1.0.0
